core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java

```diff
@@ -142,44 +152,157 @@ private Object invoke(ServiceRequestContext ctx, HttpRequest req, @Nullable Aggr
+        ParamMap paramMap = null;
+        Object[] values = new Object[parameters.size()];
+        for (int i = 0; i < parameters.size(); ++i) {
+            Parameter entry = parameters.get(i);
+            final String value;
+            switch (entry.parameterType()) {
+                case PATH_PARAM:
+                    value = ctx.pathParam(entry.name());
+                    assert value != null;
+                    values[i] = convertParameter(value, entry.type());
+                    break;
+                case PARAM:
+                    if (paramMap == null) {
+                        paramMap = paramMapOf(ctx, req, message);
+                    }
+                    value = getParam(paramMap, entry);
+                    values[i] = value != null ? convertParameter(value, entry.type()) : null;
+                    break;
+                case PREDEFINED_TYPE:
+                    if (entry.type().isAssignableFrom(ServiceRequestContext.class)) {
+                        values[i] = ctx;
+                    } else if (entry.type().isAssignableFrom(HttpRequest.class)) {
+                        values[i] = req;
+                    } else if (entry.type().isAssignableFrom(AggregatedHttpMessage.class)) {
+                        values[i] = message;
+                    } else if (entry.type().isAssignableFrom(ParamMap.class)) {
+                        if (paramMap == null) {
+                            paramMap = paramMapOf(ctx, req, message);
+                        }
+                        values[i] = paramMap;
+                    }
+                    break;
             }
         }
-        return parameters;
+        return values;
     }
 
-    BiFunction<ServiceRequestContext, HttpRequest, Object> withConverter(ResponseConverter converter) {
-        return (ctx, req) ->
-                executeSyncOrAsync(ctx, req).thenApply(obj -> convertResponse(obj, converter));
+    /**
+     * Returns a map of parameters decoded from a request.
+     */
+    private static ParamMap paramMapOf(ServiceRequestContext ctx, HttpRequest req,
+                                       @Nullable AggregatedHttpMessage message) {
+        try {
+            if (req.method() == HttpMethod.GET) {
+                return wrap(new QueryStringDecoder(ctx.query(), false).parameters());
+            }
+
+            String contentType = req.headers().get(HttpHeaderNames.CONTENT_TYPE);
+            if (contentType == null || contentType.isEmpty()) {
+                return EMPTY_MAP;
+            }
+
+            MediaType mediaType = AggregationStrategy.aggregationAvailable(contentType);
+            if (mediaType == null) {
+                return EMPTY_MAP;
+            }
+
+            assert message != null;
+            String body = message.content().toStringAscii();
+            return wrap(new QueryStringDecoder(body, mediaType.charset().orElse(ISO_8859_1), false)
```

Comment:
Complicated! 😭 
I'm not sure, but seems like there's not many frameworks following HTML5 `x-www-form-urlencoded` charset rule. [Spring framework](https://github.com/spring-projects/spring-framework/blob/master/spring-web/src/main/java/org/springframework/http/converter/FormHttpMessageConverter.java#L239) also uses `charset` from `content-type` header value, but it usually may not be specified from the client. So they may use the default charset(UTF-8) of message converter for the most cases, I think. Also, on [the decoding algorithm](https://www.w3.org/TR/html5/forms.html#application/x-www-form-urlencoded-encoding-algorithm) you mentioned, there's a note:`Parameters on the application/x-www-form-urlencoded MIME type are ignored. In particular, this MIME type does not support the charset parameter`.

I think it would be better that we apply the same rule to query string in a URI and url-encoded form data for consistency. So, how about doing as follows?

- Don't use `charset` attribute in `content-type` header value. 
- Decode the query string as UTF-8.
- If fails, decode it as ISO_8859_1 again. I'm not sure this is necessary, but I found [this from django](https://github.com/django/django/blob/master/django/http/request.py#L386).